llvm/lib/Target/AArch64/AArch64ISelLowering.cpp

```diff
@@ -6364,6 +6364,91 @@ SDValue AArch64TargetLowering::LowerINTRINSIC_WO_CHAIN(SDValue Op,
         DAG.getNode(AArch64ISD::CTTZ_ELTS, dl, MVT::i64, CttzOp);
     return DAG.getZExtOrTrunc(NewCttzElts, dl, Op.getValueType());
   }
+  case Intrinsic::experimental_vector_match: {
+    SDValue ID =
+        DAG.getTargetConstant(Intrinsic::aarch64_sve_match, dl, MVT::i64);
+
+    auto Op1 = Op.getOperand(1);
+    auto Op2 = Op.getOperand(2);
+    auto Mask = Op.getOperand(3);
+
+    EVT Op1VT = Op1.getValueType();
+    EVT Op2VT = Op2.getValueType();
+    EVT ResVT = Op.getValueType();
+
+    assert((Op1VT.getVectorElementType() == MVT::i8 ||
+            Op1VT.getVectorElementType() == MVT::i16) &&
+           "Expected 8-bit or 16-bit characters.");
+    assert(Op1VT.getVectorElementType() == Op2VT.getVectorElementType() &&
+           "Operand type mismatch.");
+    assert(!Op2VT.isScalableVector() && "Search vector cannot be scalable.");
+
+    // Note: Currently Op1 needs to be v16i8, v8i16, or the scalable versions.
+    // In the future we could support other types (e.g. v8i8).
+    assert(Op1VT.getSizeInBits().getKnownMinValue() == 128 &&
+           "Unsupported first operand type.");
+
+    // Scalable vector type used to wrap operands.
+    // A single container is enough for both operands because ultimately the
+    // operands will have to be wrapped to the same type (nxv16i8 or nxv8i16).
+    EVT OpContainerVT = Op1VT.isScalableVector()
+                            ? Op1VT
+                            : getContainerForFixedLengthVector(DAG, Op1VT);
+
+    // Wrap Op2 in a scalable register, and splat it if necessary.
+    if (Op1VT.getVectorMinNumElements() == Op2VT.getVectorNumElements()) {
+      // If Op1 and Op2 have the same number of elements we can trivially wrap
+      // Op2 in an SVE register.
+      Op2 = convertToScalableVector(DAG, OpContainerVT, Op2);
+      // If the result is scalable, we need to broadcast Op2 to a full SVE
+      // register.
+      if (ResVT.isScalableVector())
+        Op2 = DAG.getNode(AArch64ISD::DUPLANE128, dl, OpContainerVT, Op2,
+                          DAG.getTargetConstant(0, dl, MVT::i64));
+    } else {
+      // If Op1 and Op2 have different number of elements, we need to broadcast
+      // Op2. Ideally we would use a AArch64ISD::DUPLANE* node for this
+      // similarly to the above, but unfortunately we seem to be missing some
+      // patterns for this. So, in alternative, we splat Op2 through a splat of
+      // a scalable vector extract. This idiom, though a bit more verbose, is
```

Comment:
I'd rather get this fixed. Is there something I can help with?

Reply:
I believe when I tried to use DUPLANE64 nodes to achieve this for `v8i8` needle vectors, I kept facing instruction selection issues.

However, I think the main problem with the current implementation is that we have to "widen" Op2 to an SVE register before splatting it, i.e.:
```cpp
    SDValue Op2Widened = DAG.getNode(ISD::INSERT_SUBVECTOR, dl, OpContainerVT,
                                     DAG.getUNDEF(OpContainerVT), Op2,
                                     DAG.getConstant(0, dl, MVT::i64));
    Op2 = DAG.getNode(ISD::EXTRACT_VECTOR_ELT, dl, Op2IntVT,
                      DAG.getBitcast(Op2PromotedVT, Op2Widened),
                      DAG.getConstant(0, dl, MVT::i64));
    Op2 = DAG.getSplatVector(Op2PromotedVT, dl, Op2);
```

Currently, we need to do this because we seem to be missing patterns to fold an extract from a fixed vector and subsequent splat to scalable vector into a single MOV, and instead do the extract+splat explicitly (see [here](https://godbolt.org/z/1vqrdYY6a)).

Should we fix this? If so, in this patch or separately?

Reply:
Yes, this seems like a useful optimisation in its own right.  Definitely a separate patch.  Let me know if you want to run with it, otherwise I'm happy to put a PR up on Monday.

Reply:
I don't think I can do it until Monday, so if you want to take it that would be much appreciated, otherwise I'll do so later next week.

Reply:
FYI: I got a little sidetracked but will put a PR up for this tomorrow.

Reply:
Sounds good, thank you very much. If you'd like me to have a look instead just let me know!